core/arch/arm/kernel/thread_optee_smc.c

```diff
@@ -523,20 +530,26 @@ static struct mobj *get_rpc_alloc_res(struct optee_msg_arg *arg,
 {
 	struct mobj *mobj = NULL;
 	uint64_t cookie = 0;
+	size_t sz = 0;
+	paddr_t p = 0;
 
 	if (arg->ret || arg->num_params != 1)
 		return NULL;
 
 	if (arg->params[0].attr == OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT) {
+		p = arg->params[0].u.tmem.buf_ptr;
+		sz = arg->params[0].u.tmem.size;
 		cookie = arg->params[0].u.tmem.shm_ref;
-		mobj = mobj_shm_alloc(arg->params[0].u.tmem.buf_ptr,
-				      arg->params[0].u.tmem.size,
-				      cookie);
+
+		if (core_pbuf_is(CORE_MEM_NSEC_SHM, p, sz))
+			mobj = mobj_shm_alloc(p, sz, cookie);
+		else if (IS_ENABLED(CFG_CORE_DYN_SHM) && sz <= SMALL_PAGE_SIZE)
+			mobj = rpc_shm_mobj_page_alloc(p, cookie);
```

Comment:
I think the 2 occurrences (here and line 312) could be factorize into a single function, not just the part in `rpc_shm_mobj_page_alloc()`:
```c
static struct mobj *rpc_shm_mobj_alloc(paddr_t pa, size_t sz, uint64_t cookie)
{
	if (core_pbuf_is(CORE_MEM_NSEC_SHM, pa, sz))
		return mobj_shm_alloc(pa, sz, cookie);

	if (IS_ENABLED(CFG_CORE_DYN_SHM) &&
	    !(pa & SMALL_PAGE_MASK) && sz <= SMALL_PAGE_SIZE)
		return mobj_mapped_shm_alloc(&pa, 1, 0, cookie);

	return NULL;
}